docs/cudf/source/developer_guide/benchmarking.md

```diff
@@ -0,0 +1,204 @@
+_unless_ developers expect certain arguments to trigger code paths with very different performance characteristics.
+
+```{note}
+`pytest` does not support having two benchmark files with the same name, even if they are in separate directories.
+Therefore, benchmarks of internal methods of _public_ classes go in files suffixed with `_internal`.
+Benchmarks of `DataFrame._apply_boolean_mask`, for instance, belong in `internal/bench_dataframe_internal.py`.
+```
+
+Free functions have more flexibility.
+Broadly speaking, they should be grouped into benchmark files containing similar functionality.
+For example, I/O benchmarks can all live in `bench_io.py`.
+For now those groupings are left to the discretion of developers.
+
+## Standard fixtures
+
+The most difficult part of writing a benchmark is often deciding what class of objects that benchmark should run for.
+Developers often need to run the same benchmarks with _multiple_ `pytest` fixtures rather than just one.
+For instance, benchmarks in `API/bench_indexedframe.py` must be run for both `DataFrame` and `Series`.
+To address this requirement, we make use of [`pytest_cases`](https://smarie.github.io/python-pytest-cases/).
+Specifically we use `pytest_cases.fixture_union` to create fixture unions.
+Tests parametrized by fixture unions are run once for each member of the union.
+
+This feature is critical to the design of our benchmarks.
+It allows developers to write a single benchmark `def bench_foo(indexedframe)`,
+and then have that benchmark automatically run many times with different parameters.
+In particular, our standard fixtures cover the following:
+
+- Class: Objects of a specific class, e.g. `DataFrame`.
+- Nullability: Objects with and without null entries.
+- Dtype: Objects of a specific dtype.
+- Size: Objects with a certain number of rows or columns.
+
+Fixture names identify these parameters unambiguously like so:
+`{classname}_dtype_{dtype}[_nulls_{true|false}][[_cols_{num_cols}]_rows_{num_rows}]`.
+If a fixture name does not contain a particular component, it represents a union of all values of that component.
+For example, a benchmark
+```
+def bench_foo(benchmark, dataframe_dtype_int_rows_100):
+    benchmark(df.foo)
+```
+will run for both nullable and non-nullable 100 row integer `DataFrame`s of with different numbers of columns.
+
+These fixtures should support most use cases.
+Developers may define custom fixtures if necessary, but this should be done with care.
+The default fixtures provide reasonable benchmark coverage without excessive resource usage.
+More bespoke fixtures, if necessary, should be constructed with the same constraints in mind.
+Furthermore, the default fixtures are designed to work when
+[comparing to pandas](pandascompare) or [running tests](testing).
+New fixtures must also account for these use cases.
+
+### The `accepts_cudf_fixture` decorator
```

Comment:
Hmm I'm not sure that we want to abstract away the fact that a fixture is being used though. We're trying to hide the mapping between fixtures and parameters, but we expect users to be familiar with pytest and write their tests accordingly, right? cf. [this conversation on the benchmarking thread](https://github.com/rapidsai/cudf/pull/11125#discussion_r903088075), it seems relevant to how we want devs to "think" about these benchmarks.

Reply:
Do developers have to learn fixtures in order to make use of this decorator? From the current design, it seems like it has abstracted enough detail so that it's less important whether you mention fixture in the name of the decorator or not. New contributor would look at the decorator:
```python
@accept_cudf_fixture(cls="column, ...)
def bench_col(benchmark, column):
...
```
and learn that some parameterized column will be used to parameterize the test. And that suffices his/her needs. They are certainly welcome to read in depth about its design but I don't think that's not most contributors would do.

Reply:
It sounds like both you and @shwina are in favor of a rename, which I support. What name do you think makes sense? `with_frame` is too specific since the object could also be a column. `with_object` seems too generic. What else would you suggest?

Reply:
- `benchmark_target`?
- `with_cudf_object`?

Reply:
After discussing offline, we settled on `benchmark_with_object`.